library/index.c

```diff
@@ -352,27 +353,43 @@ static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
 		if (num > b->cfg.index_block_size)
 			num = b->cfg.index_block_size;
 
-		search_start = bctl->sort.data + block->start_offset;
 		/*
 		 * We do not use @block->end_offset here, since it points to
 		 * the start offset of the *next* record, which potentially
 		 * can be outside of the index, i.e. be equal to the size of
 		 * the index.
 		 */
-		search_end = search_start + (num - 1);
+		hdr_block_size = num * hdr_size;
+		hdr_block_offset = block->start_offset;
 	} else {
 		pthread_rwlock_unlock(&bctl->index_blocks_lock);
-		goto out;
+		goto err_out_exit;
 	}
 	pthread_rwlock_unlock(&bctl->index_blocks_lock);
 
 	st->bsearch_reached++;
 
+	hdr_block = malloc(hdr_block_size);
+	if (!hdr_block) {
+		err = -ENOMEM;
+		goto err_out_exit;
+	}
+
+	bytes = pread(bctl->sort.fd, hdr_block, hdr_block_size, hdr_block_offset);
+	if (bytes != hdr_block_size) {
+		if (bytes < 0)
+			err = -errno;
+		goto err_out_free_index;
+	}
```

Comment:
Again we have only one chance to read whole data. Why doesn't it use `__eblob_read_ll` ?